aidial_adapter_bedrock/dial_api/resource.py

```diff
@@ -0,0 +1,180 @@
+                f"Can't derive content type of the {self.entity_name}"
+            )
+
+        if (
+            self.supported_types is not None
+            and type not in self.supported_types
+        ):
+            raise UnsupportedContentType(
+                message=f"The {self.entity_name} is not one of the supported types",
+                type=type,
+                supported_types=self.supported_types,
+            )
+
+        return type
+
+
+class URLResource(DialResource):
+    url: str
+    content_type: str | None = None
+
+    @root_validator
+    def validator(cls, values):
+        values["entity_name"] = values.get("entity_name") or "URL"
+        return values
+
+    async def download(self, storage: FileStorage | None) -> Resource:
+        type = await self.get_content_type()
+        data = await _download_url(storage, self.url)
+        return Resource(type=type, data=data)
+
+    async def guess_content_type(self) -> str | None:
+        return (
+            self.content_type
+            or Resource.parse_data_url_content_type(self.url)
+            or mimetypes.guess_type(self.url)[0]
+        )
+
+    def is_data_url(self) -> bool:
+        return Resource.parse_data_url_content_type(self.url) is not None
+
+    async def get_resource_name(self, storage: FileStorage | None) -> str:
+        if self.is_data_url():
+            return f"data URL ({await self.guess_content_type()})"
+
+        name = self.url
+        if storage is not None:
+            name = await storage.get_human_readable_name(self.url)
+
+        return truncate_string(name, n=50)
+
+
+class AttachmentResource(DialResource):
+    attachment: Attachment
+
+    @validator("attachment", pre=True)
+    def parse_attachment(cls, value):
+        if isinstance(value, dict):
+            attachment = Attachment.parse_obj(value)
+            # Working around the issue of defaulting missing type to a markdown:
+            if "type" not in value:
+                attachment.type = None
+            return attachment
+        return value
+
+    @root_validator(pre=True)
+    def validator(cls, values):
+        values["entity_name"] = values.get("entity_name") or "attachment"
+        return values
+
+    async def download(self, storage: FileStorage | None) -> Resource:
+        type = await self.get_content_type()
+
+        if self.attachment.data:
+            data = base64.b64decode(self.attachment.data)
+        elif self.attachment.url:
+            data = await _download_url(storage, self.attachment.url)
+        else:
+            raise ValidationError(f"Invalid {self.entity_name}")
+
+        return Resource(type=type, data=data)
+
+    def create_url_resource(self, url: str) -> URLResource:
+        return URLResource(
+            url=url,
+            content_type=self.effective_content_type,
+            entity_name=self.entity_name,
+        )
+
+    @property
+    def effective_content_type(self) -> str | None:
```

Comment:
What do you mean by `effective`?

Reply:
Effective means informative. Catch-all binary content type is not informative, so it's discounted in this method.

Reply:
Maybe let's rename it to `informative_content_type` then?

Or at least leave some comment about this

Because `effective` adjective in terms of content type may assume something about encoding effectiveness itself, which can be confusing

Reply:
Or we can use `meaningful_content_type` for that, what do you think?

Reply:
Renamed to `informative_content_type`